db/db_impl.cc

```diff
@@ -3772,8 +3776,8 @@ size_t DBImpl::GetWalPreallocateBlockSize(uint64_t write_buffer_size) const {
   size_t bsize = write_buffer_size / 10 + write_buffer_size;
```

Comment:
add `mutex_.AssertHeld()` and make sure all caller lock the mutex. You can do something like

  size_t preallocate_block_size = GetWalPreallocateBlockSize();
  mutex_.Unlock();
  ...
  // use preallocate_block_size instead of calling GetWalPreallocateBlockSize().